x/application/keeper/prune_undelegations.go

```diff
@@ -0,0 +1,50 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sessionkeeper "github.com/pokt-network/poktroll/x/session/keeper"
+)
+
+// NumSessionsAppToGatewayUndelegationRetention is the number of sessions for which
+// undelegation from applications to gateways are delayed before being pruned.
+// TODO_DOCUMENT(@red-0ne): Need to document the flow from this comment
+// so its clear to everyone why this is necessary; https://github.com/pokt-network/poktroll/issues/476#issuecomment-2052639906.
+// TODO_CONSIDERATION(#516): Should this be configurable? Note that it should
+// likely be a function of SubmitProofCloseWindowNumBlocks once implemented.
+const NumSessionsAppToGatewayUndelegationRetention = 2
+
+// EndBlockerPruneAppToGatewayPendingUndelegation runs at the end of each block
+// and prunes app to gateway undelegations that have exceeded the retention delay.
+func (k Keeper) EndBlockerPruneAppToGatewayPendingUndelegation(ctx sdk.Context) error {
+	currentHeight := ctx.BlockHeight()
+
+	// Calculate the block height at which undelegations should be pruned
+	numBlocksUndelegationRetention := GetNumBlocksUndelegationRetention()
+	if currentHeight <= numBlocksUndelegationRetention {
+		return nil
+	}
+	pruningBlockHeight := uint64(currentHeight - numBlocksUndelegationRetention)
```

Comment:
The name confused me for a minute here. I was expecting it to be used with `<=` below. It seems to me it may be better described as `earliestUndelegationRetentionHeight` or `firstUnprunedUndelegationHeight` or something like that.